### PR TITLE
KAFKA-10892: Shared Readonly State Stores

### PR DESCRIPTION
**Pull request by calmera**
This PR implements KIP-813: support for shareable state stores.

Tests were created to validate the structure of the resulting topology. 

Certain parts (MockProcessor eg.) have been migrated to the newer PAPI as well. Tests have been ran against the whole project to be sure this migration doesn't cause any side-effects.

The contribution is my original work and I license the work to the project under the project's open source license

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 1]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
         return this;
     }
 
+    /**
+     * Adds a Read Only {@link StateStore} to the topology.
+     *
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java:746]: nit: do we need a `<p>` tag here?

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 2]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
         return this;
     }
 
+    /**
+     * Adds a Read Only {@link StateStore} to the topology.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: nit: `read-only`

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 3]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     *
+     * A Read Only StateStore can use any compacted topic as a changelog.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: Proposal:
```
A read-only state store uses its input topic for fault-tolerance. Thus, in contrast to regular state stores, it must never create an internal changelog topic. Therefore, the input topic should be configured with log compaction.
```

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 4]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     * <p>
+     * A {@link SourceNode} will be added to consume the data arriving from the partitions of the input topic.
+     * <p>
+     * The provided {@link ProcessorSupplier} will be used to create an {@link ProcessorNode} that will receive all
+     * records forwarded from the {@link SourceNode}.
+     * This {@link ProcessorNode} should be used to keep the {@link StateStore} up-to-date.
+     *
+     * @param storeBuilder          user defined key value store builder
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: If we are limited to kv-store, should we change the type to `StoreBuilder<KeyValueStore>` (or similar)?

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 5]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+                                                                  final String sourceName,
+                                                                  final TimestampExtractor timestampExtractor,
+                                                                  final Deserializer<KIn> keyDeserializer,
+                                                                  final Deserializer<VIn> valueDeserializer,
+                                                                  final String topic,
+                                                                  final String processorName,
+                                                                  final ProcessorSupplier<KIn, VIn, Void, Void> stateUpdateSupplier) {
+        if (storeBuilder.loggingEnabled()) {
+            // -- disabling logging. We might want to print some logging.
+            storeBuilder.withLoggingDisabled();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: I think we should throw a `TopologyException` here (we do the same for global stores if logging is enabled).

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamBranchTest.java [hunk 6]
```diff
@@ -59,7 +59,7 @@ public void testKStreamBranch() {
 
         assertEquals(3, branches.length);
 
-        final MockProcessorSupplier<Integer, String> supplier = new MockProcessorSupplier<>();
+        final MockProcessorSupplier<Integer, String, Void, Void> supplier = new MockProcessorSupplier<>();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamBranchTest.java:62]: Seems this changes are not related to this PR (similar below)? Would be good to exclude them and put into it's own PR.

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. And sorry for the long wait on a review.

As the store should be read-only, we should extend `ProcessorContext#getStateStore` to return a read only store if any other Processor get a handle on the store. We do a similar thing for global stores to protect them. I guess, the tricky thing will be to know if a store is read-only -- maybe we need to track some metadata about it.

Can you also rebase the PR to resolve the conflicts. Thx.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 4 above)
Comment (calmera) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java]: There is no real restriction on the type of store, so will adapt the description.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamBranchTest.java (hunk 6 above)
Comment (calmera) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamBranchTest.java:62]: Yeah, problem is that I need them to decently test the topology.

Let me see what I can do

**Comment by calmera** [comment]
Split this PR into two new ones:

- #12740 updating the PAPI test classes to the latest version
- #12742 handling the Topology changes and read-only statestore introduction
